Notification: list different types + more information about webhooks (#31)
packagist/private-packagist-docs: docs/notifications.md

# Notifications

### Notification Types

#### Package Releases

> Private Packagist will notify you every time we find one or more new releases for one of your packages.

Reviewer: Can you explain what "releases" are? It should really say "versions" here and explain that tags, branches, commits are new or modified versions?

Author: done

> Private Packagist will notify you every time we find one or more new versions for one of your packages. All tags, branches and commits with a composer.json file in the root directory are either new or modified versions.

Reviewer: This sentence doesn't really make sense to me, "Tags [...] are either new or modified versions.", pretty sure that's only for new tags?

Author: If the source reference of a tag changes then we do update the version in Private Packagist and this will trigger a release notification for the tag.

Reviewer: @glaubinix Sure then "New or changed tags". But a tag is not constantly a new or modified version, that sentence is just wrong.

> Notification channels allow you to specify about which private packages you want to get notified and whether you also want to receive notifications for mirrored packages as well.

Reviewer: This is a bit out of the blue, what's a notification channel? At least mention they are explained below?

Author: Remove `notification channels` to not confuse users when reading this

> Additionally, you can filter for releases by minimum stability: "dev / any" will match any release including commits to a branch whereas "stable" will only match releases considered stable releases by Composer.

Reviewer: In addition to what? Just remove the word "Additionally"?

Reviewer: What releases does Composer consider stable? Maybe add a link to composer docs if you don't want to explain all the options in detail here?

Author: Done

#### Security Alerts

> As part of our [security monitoring](./security-monitoring.md) feature, you can configure notification channels to receive a notification every time one or more security issues are found for one of your projects.

Reviewer: ```suggestion
[Security Monitoring](./security-monitoring.md) allows you to select notification channels to receive a notification every time security issues are found in dependencies of one of your projects.
```

#### Security Summaries

> In addition to security alerts your notification channels can also receive either weekly or monthly summaries including information about all open security issues in your organization.

Reviewer: ```suggestion
In addition to immediate security alerts your notification channels can also receive either weekly or monthly summaries listing all open security issues in dependencies of your organization's projects.
```

### Configuring Notifications
Every user receives security notifications by email for all projects they have access to by default.

> Users can unsubscribe either from individual projects or from all security notifications if they do not wish to receive any email notifications.

Reviewer: ```suggestion
Users can unsubscribe either from individual projects or from all security notifications if they do not wish to receive email notifications.
```

Notification channels allow you to receive notifications via other means than email to user accounts. The following types of notification channels are available:
- **Email**: Sends notifications to a list of email addresses
- **Slack Webhook**: Sends notifications to your configured Slack channel
- **Microsoft Teams Webhook**: Sends notifications to your configured Microsoft Teams channel
- **Webhook**: Sends an HTTP POST request to a defined URL optionally signed with a user supplied secret.

Notification channels can be added on your organization’s settings page under *Notification Channels -> Add Notification Channel*.

### Receiving Webhook Notifications

Webhook notifications are sent as HTTP POST requests to the endpoint configured with the notification event data send as payload.

> Client, server, and network errors will automatically be retried up to five times.

Author: 4XX errors can be caused by various different reasons and might because by an invalid setup e.g. 404 because wrong URL posted or 403/401 authentication missing/changed/not setup.

> Client, server, and network errors will automatically be retried up to five times.

Reviewer: @glaubinix then let's not call them client errors, http errors?

#### Delivery Headers

HTTP POST payloads that are delivered to your webhook's configured URL endpoint will contain several special headers:

<table>
    <thead>
        <tr>
            <th>Header</th>
            <th>Description</th>
        </tr>
    </thead>
    <tbody>
        <tr>
            <td>Packagist-Event</td>
            <td>Name of the notification event</td>
        </tr>
        <tr>
            <td>Packagist-Notification</td>
            <td>Notification identifier, allows you to track a notification across multiple retries</td>
        </tr>
        <tr>
            <td>Packagist-Delivery</td>
            <td>Identifier for the current request/delivery</td>
        </tr>
        <tr>
            <td>Packagist-Signature</td>
            <td>Signature to validate the request based on the payload. This header will only be sent if a webhook secret is configured.</td>
        </tr>
    </tbody>
</table>

#### Webhook Request Validation

> We do recommend that you set up a webhook secret and validate the payload either using [our api client](https://github.com/packagist/private-packagist-api-client#validate-incoming-webhook-payloads) or by running `hash_equals('sha1='.hash_hmac('sha1', (string) $request->getBody(), $SECRET_USER_CHOSEN), $response->getHeader('Packagist-Signature'));`

Reviewer: ```suggestion
We recommend that you set up a webhook secret and validate the payload either using [our api client](https://github.com/packagist/private-packagist-api-client#validate-incoming-webhook-payloads) or by running `hash_equals('sha1='.hash_hmac('sha1', (string) $request->getBody(), $SECRET_USER_CHOSEN), $response->getHeader('Packagist-Signature'));`
```
